File: sol/solution/0101-0200/0109/Solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {ListNode} head
# @return {TreeNode}
def sorted_list_to_bst(head)
    return nil if head.nil?
    return TreeNode.new(head.val) if head.next.nil?
    slow = head
    fast = head
    prev = nil
    while fast && fast.next
        prev = slow
        slow = slow.next
        fast = fast.next.next
    end
    prev.next = nil if prev && prev != slow
    root = TreeNode.new(slow.val)
    root.left = sorted_list_to_bst(prev.nil? ? nil : head)
    root.right = sorted_list_to_bst(slow.next)
    root
end

File: sol/solution/0101-0200/0109/Solution.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    /**
     * @param ListNode $head
     * @return TreeNode
     */
    function sortedListToBST($head) {
        if (!$head) return null;
        if (!$head->next) return new TreeNode($head->val);
        $slow = $fast = $head;
        $prev = null;
        while ($fast && $fast->next) {
            $prev = $slow;
            $slow = $slow->next;
            $fast = $fast->next->next;
        }
        if ($prev) $prev->next = null;
        $root = new TreeNode($slow->val);
        $root->left = $this->sortedListToBST($head === $slow ? null : $head);
        $root->right = $this->sortedListToBST($slow->next);
        return $root;
    }
}